packages/wrangler/src/deployment-bundle/source-maps.ts

```diff
@@ -0,0 +1,136 @@
+		return scanSourceMaps([main, ...modules]);
+	}
+}
+
+interface SourceMapV3 {
+	version: 3;
+	file: string;
+	sourceRoot: string;
+	sources: string[];
+	sourcesContent: string[];
+	mappings: string;
+	names: string[];
+}
+
+/**
+ * Load and normalize a source map emitted by Wrangler using the given path and
+ * directory metadata.
+ */
+function loadSourceMap(
+	{ name, filePath }: CfModule,
+	sourceMapPath: string,
+	{ entryDirectory }: SourceMapMetadata
+): CfWorkerSourceMap[] {
+	if (filePath === undefined) {
+		return [];
+	}
+	const map = JSON.parse(
+		fs.readFileSync(sourceMapPath).toString()
+	) as SourceMapV3;
+	// Overwrite the file property of the source map to match the name of the
+	// main module in the multipart upload.
+	map.file = name;
+	// Remove the temporary directory prefix generated by Wrangler that appears
+	// in the source root path.
+	const sourceRootPath = path.dirname(path.join(entryDirectory, sourceMapPath));
+	map.sourceRoot = stripPrefix(sourceRootPath, map.sourceRoot);
+	map.sources = map.sources.map((source) => {
+		const originalPath = path.join(path.dirname(filePath), source);
+		// Remove the temporary build directory prefix generated by Wrangler
+		// that appears in the source path.
+		return stripPrefix(withTrailingSep(entryDirectory), originalPath);
+	});
+	return [
+		{
+			name: name + ".map",
+			content: JSON.stringify(map),
+		},
+	];
+}
+
+/**
+ * Find source maps by scanning module contents for special `//#
+ * sourceMappingURL=` comments pointing to source map files.
+ */
+function scanSourceMaps(modules: CfModule[]): CfWorkerSourceMap[] {
+	const maps: CfWorkerSourceMap[] = [];
+	for (const module of modules) {
+		const commentPrefix = "//# sourceMappingURL=";
+		const lastLine = module.content.toString().split("\n").pop();
+		if (
+			lastLine === undefined ||
+			!lastLine.startsWith(commentPrefix) ||
+			module.filePath === undefined
+		) {
+			continue;
+		}
+		// Assume the source map path in the comment is relative to the
+		// generated file it appears in.
+		const commentPath = stripPrefix(commentPrefix, lastLine).trim();
+		if (commentPath.startsWith("data:")) {
+			throw new Error(
+				`Invalid source map path in ${module.filePath}: expected file path but found data URL.`
+			);
+		}
+		// Convert source map path to an absolute path that we can read.
+		const wranglerPath = path.join(path.dirname(module.filePath), commentPath);
+		if (!fs.existsSync(wranglerPath)) {
+			throw new Error(
+				`Invalid source map path in ${module.filePath}: ${wranglerPath} does not exist.`
+			);
+		}
+		const map = JSON.parse(
+			fs.readFileSync(wranglerPath).toString()
+		) as SourceMapV3;
+		// Overwrite the file property of the sourcemap to match the name of the
+		// corresponding module in the multipart upload.
+		map.file = module.name;
+		map.sourceRoot = cleanPathPrefix(map.sourceRoot);
+		map.sources = map.sources.map(cleanPathPrefix);
+		maps.push({
+			name: module.name + ".map",
```

Comment:
If `module.name + ".map"` does not match `commentPath` then the downstream tooling might be unable to link the source to this source map.

Reply:
By "downstream tooling" do you mean the runtime? We ignore both the `//# sourceMappingURL=` comment and the source map module name, and instead match the JS module name with the `file` property of the source map (which this PR overwrites to match the corresponding JS module name).

Reply:
OK, that is slightly strange behaviour from the runtime.
But I assume that in the future there will be other consumers that will read the source file and then request the source map? Like quick editor dashboard?
Or is there going to be some special API for these consumers?

Reply:
The Worker source and source map are downloaded using the same endpoint and can be downloaded together in a single request. If clients want to correlate source files and source maps they can match them using the same logic that the runtime does (JS module name and source map `file` field). Or at least that was the plan.

Reply:
Yeah but many SourceMap consumers will parse the source file and use the path in the comment if it is there.
They only fall back to appending the `.map` extension if there is no comment.
It might be better to strip out the comment if you don't expect it to be used.

Not a big issue but might be worth keeping an eye on as this starts to be used.